File: Seori/LeetCode 75/Graph_DFS/841. Keys and Rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        
        def dfs(now: int, rooms: List[List[int]]) -> None:
            visited[now] = 1
            for next in rooms[now]:
                if next < n and not visited[next]:
                    dfs(next, rooms)
        
        # [1] DFS로 0번방부터 탐색
        n = len(rooms)
        visited = [0] * n
        dfs(0, rooms)

        # [2] 모든 방에 방문했다면 True
        return sum(visited) == n

File: Gombang/[LeetCode75] - [DFS]/841_Keys_and_Rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {

        vector<bool> visited(rooms.size(), false);
        queue<int> queue;

        // 큐 초기화
        const int FIRST_ROOM = 0;
        visited[FIRST_ROOM] = true;
        for (auto& keyNumber : rooms[FIRST_ROOM])
        {
            queue.push(keyNumber);
        }

        while (queue.size() > 0)
        {
            int key = queue.front();
            queue.pop();

            if (visited[key])
                continue;

            visited[key] = true;

            for (auto& keyNumber : rooms[key])
            {
                queue.push(keyNumber);
            }
        }

        for (int i = 0; i < visited.size(); i++)
        {
            if (visited[i] == false)
                return false;
        }

        return true;
    }
};